### ISSUE TITLE
vector/matrix reference within double dot notation

### ISSUE THREAD

**Issue by kgoldfeld** [kgoldfeld/simstudy#102]
It would be nice to be able to refer to vectors and matrices in a dynamic way in a formula that uses double do notation. I think this is easiest to see in a simple example. I have two vectors of length 2 (`b_1` and `b_2`) and a `2 X 2`  matrix `b_12`:

```r
library(simstudy)

b_0 <- 0
b_1 <- c(-3, 3)
b_2 <- c(-1, 1)
b_12 <- matrix(c(-1, -1, -1, 3), nrow = 2)
```

Each individual is assigned a value of 1 or 2 for factor `a` and for factor `b`:
```r
d1 <- defData(varname = "a", formula = ".5;.5", variance = "1;2", dist = "categorical")
d1 <- defData(d1, varname = "b", formula = ".5;.5", variance = "1;2", dist = "categorical")

set.seed(1)
dx <- genData(3, d1)
dx
#>    id a b
#> 1:  1 2 1
#> 2:  2 2 2
#> 3:  3 1 1
```

Now, I want to create an outcome variable that references these vectors and matrix, and pulls from the cell based on the values of `a` and `b`. Here is a non-simstudy solution:

```r
dx[, y := b_0 + b_1[a] + b_2[b] + b_12[a, b], keyby = id]
dx
#>    id a b  y
#> 1:  1 2 1  1
#> 2:  2 2 2  7
#> 3:  3 1 1 -5
```

It would be nice to be able to use double dot notation to refer to the vectors and matrix from within the the dataDef table, like this:

```r
d1 <- defData(d1, varname = "y", formula = "..b_0 + ..b_1[a] + ..b_2[b] + ..b_12[a, b]", dist="nonrandom")
d1
#>    varname                                    formula variance        dist         link
#> 1:       a                                      .5;.5      1;2 categorical     identity
#> 2:       b                                      .5;.5      1;2 categorical     identity
#> 3:       y ..b_0 + ..b_1[a] + ..b_2[b] + ..b_12[a, b]        0   nonrandom     identity
```

It doesn't quite work:

```r
set.seed(1)
genData(3, d1)
#>    id a b  y
#> 1:  1 2 1  1
#> 2:  1 2 1  5
#> 3:  1 2 1  1
#> 4:  2 2 2  3
#> 5:  2 2 2  7
#> 6:  2 2 2  3
#> 7:  3 1 1 -5
#> 8:  3 1 1 -5
#> 9:  3 1 1 -5
```

**Comment by assignUser** [comment]
It triples the rows... interesting. I'll look into it.

**Comment by assignUser** [comment]
The problem is with how r handles the subsetting and addition of the elements. Vectorizing the formula works for this example but is not as easily generalized. We will probably need to detect non-integer indices etc... currently the formula parsing is pretty barebones, I'll think about how to do this,
```r
b_0 + b_1[a] + b_2[b] + b_12[a, b]
#      [,1] [,2] [,3]
# [1,]    1    5    1
# [2,]    3    7    3
# [3,]   -5   -5   -5

f <- function(a,b){b_0 + b_1[a] + b_2[b] + b_12[a, b]}
ff <- Vectorize(f)
ff(a,b)
# [1]  1  7 -5
```

**Comment by assignUser** [comment]
The "easiest" way seems to be to actually use data.table to evaluate the formulas as you did in your example. That saves a lot of work on regex, string work to detect matrices and handle the indices differently etc.. 
I have created a new branch for this issue where I have implemented this but it is not final as test are failing all over the place (as now the dt is no longer optional / needs other handling, but no serious issues.). And it might have performance implication which I need to check to.

Here are the changes that work for your example!
https://github.com/kgoldfeld/simstudy/blob/636b297754bbf19e7b76c0b8a363b8429ad5b366/R/internal_utility.R#L78-L84
